### DISCUSSION TITLE
Using object parameters in NED

### DISCUSSION THREAD

**Discussion by StephanOostveen** [omnetpp/omnetpp#986] [Q&A] [answered]
Hi,

I am trying to use the [object parameters/structured data](https://doc.omnetpp.org/omnetpp/manual/#sec:ned-lang:object-parameters) feature of omnet 6 to define some message structure. During network setup I get the following error:

> Cannot assign parapmeter ' canInput{...}: Class "CanList" not found -- perhaps its code was not linked in, or the class was not registered with Register_Class(), or in the case of modules and schannels with Define_Module()/Define_Channel()

Similar to the Trail example in the simulation manual (section 3.6.8) I used the following message file to define the classes:

```
namespace FiCo4OMNeT;

class CanDataFrameDefinition extends cObject {
    unsigned int canID; // ID of the message
    unsigned int sizeInBytes;
    string bus;
}

class CanList extends cObject {
    CanDataFrameDefinition definition[];
}
```

Which I wish to use in the following simple module (code reduced for brevity):
```
package fico4omnet.operatingsystem;

import fico4omnet.operatingsystem.CanList;

simple Logical {
    parameters:
        object canInput = default(CanList{messages: []});
        object canOutput = default(CanList{messages: []});
}
```

The logical is then imported and used inside a compound module and part of a network.

I have checked that the C++ code for the messages is generated and included correctly as I can use their definition in other C++ code. It seems that I am missing the correct way to include the message definition in the ```Logical``` ned definition.

Kind regards,
Stephan

**Comment by rhornig** [comment]
As a first sanity check, see if the generated `_m.cc` file contains the proper `Register_Class()` macros (it should, unless there is some serious bug in the message generator).

The usual cause of these messages is that class in question is not actually linked into the final executable. The list of object files to be linked are generated at the time when you run `opp_makemake`. It is a typical error that you add a new `.cc` or `.msg` file to your project and you forget to re-generate the project makefile with `opp_makemake`. If you forget this, the given class will indeed be missing. You can also check manually if the class is present in the registraion list in Qtenv: From the main menu select `Inspect | Available Components | Registered Classes`. The `CanList` class **must** be present in that list. If it is not, either the `Register_Class()` macro call is missing or the `.cc` file is not included at all for the linker.

**Reply by StephanOostveen** [reply]
Thanks for the quick reply

Maybe good to mention, I am using the [Omnetpp/CMake](https://github.com/omnetpp/cmake) project as a build environment. I have added the message files successfully using the ```generate_opp_message``` cmake function and have validated that it is both generated, the generated code is compiled and linked into the final library by checking the output of  ```ninja -v```

The generated _m.cc file does contain the Register_Class() macros, both for the CanList and CanDataFrameDefinition classes, together with calls to Register_ClassDescriptor() for the associated descriptor classes. 

I checked the registration list in Qtenv, nice tip I didn't know this feature yet, and the ```CanList``` class is indeed present in that list.

**Reply by rhornig** [reply, marked as answer]
I would like to point out that your message file generates the class into the `FiCo4OMNeT` namespace while the error message complains about missing the class in the global namespace. My guess is that this is a namespace mismatch.

**Reply by StephanOostveen** [reply]
Oh good catch thank you, prepending the namespace in the object parameter definition in ```Logical.ned``` as follows ```object canInput = default(FiCo4OMNeT: :CanList{messages: []});``` fixes the error. 

Unfortunately a new error occurred: 
```
Cannot assign parameter 'caninput': Object parameters may only hold objects they fully own but object is 
not a cOwnedObject and does not keep track of its owner
```

Changing the CanList message base class from ```cObject``` to ```cOwnedObject``` seems to let the error disappear. But the error message also mentioned using the ```dup()``` function as an alternative fix. I tried searching the simulation manual for a reason why either ```cOwnedObject``` or ```dup()``` would be necessary in this case but couldn't find a satisfying reason, could you help me understand the decision for the ```cOwnedObject``` base class in this case? Also it would help changing the [relevant section](https://doc.omnetpp.org/omnetpp/manual/#sec:ned-lang:object-parameters) in the simulation manual to inherit from the right base class and maybe provide a small explanation.

**Reply by rhornig** [reply]
Objects can reference other objects (even in a circular way) and the simulation kernel MUST know who is responsible for deleting the object if it's no longer needed. The only way to enforce this is to actually keep track of the owner (by using the cOwnedObject as a base class).

**Reply by StephanOostveen** [reply]
Ah oke that makes sens thanks for the help :)
